python/ccao/vars_funs.py

```diff
@@ -126,3 +126,155 @@ def vars_rename(
         # If the input data is a list, it's not possible to update it inplace,
         # so ignore that argument
         return [mapping.get(col, col) for col in data]
+
+
+def vars_recode(
+    data: pd.DataFrame,
+    cols: list[str] | None = None,
+    code_type: str = "long",
+    as_factor: bool = True,
+    dictionary: pd.DataFrame | None = None,
+) -> pd.DataFrame:
+    """
+    Replace numerically coded variables with human-readable values.
+
+    The system of record stores characteristic values in a numerically encoded
+    format. This function can be used to translate those values into a
+    human-readable format. For example, EXT_WALL = 2 will become
+    EXT_WALL = "Frame + Masonry". Note that the values and their translations
+    must be specified via a user-defined dictionary. The default dictionary is
+    :data:`vars_dict`.
+
+    Options for ``code_type`` are:
+
+    - ``"long"``, which transforms EXT_WALL = 1 to EXT_WALL = Frame
+    - ``"short"``, which transforms EXT_WALL = 1 to EXT_WALL = FRME
+    - ``"code"``, which keeps the original values (useful for removing
+      improperly coded values, see the note below)
+
+    :param data:
+        A pandas DataFrame with columns to have values replaced.
+    :type data: pandas.DataFrame
+
+    :param cols:
+        A list of column names to be transformed, or ``None`` to select all columns.
+    :type cols: list[str]
+
+    :param code_type:
+        The recoding type. See description above for options.
+    :type code_type: str
+
+    :param as_factor:
+        If True, re-encoded values will be returned as categorical variables
+        (pandas Categorical).
+        If False, re-encoded values will be returned as plain strings.
+    :type as_factor: bool
+
+    :param dictionary:
+        A pandas DataFrame representing the dictionary used to translate
+        encodings.
+    :type dictionary: pandas.DataFrame
+
+    :raises ValueError:
+        If the dictionary is missing required columns or if invalid input is
+        provided.
+
+    :return:
+        The input DataFrame with re-encoded values for the specified columns.
+    :rtype: pandas.DataFrame
+
+    .. note::
+        Values which are in the data but are NOT in the dictionary will be
+        converted to NaN.
+
+    :example:
+
+    .. code-block:: python
+
+        import ccao
+
+        sample_data = ccao.sample_athena
+
+        # Defaults to `long` code type
+        ccao.vars_recode(data=sample_data)
+
+        # Recode to `short` code type
+        ccao.vars_recode(data=sample_data, code_type="short")
+
+        # Recode only specified columns
+        ccao.vars_recode(data=sample_data, cols="GAR1_SIZE")
+    """
+    # Validate the dictionary schema
+    dictionary = dictionary if dictionary is not None else vars_dict
+    if dictionary.empty:
+        raise ValueError("dictionary must be a non-empty pandas DataFrame")
+
+    required_columns = {"var_code", "var_value", "var_value_short"}
```

Comment:
**issue (blocking):** `var_name`, `var_type`, and `var_data_type` should be required as well since they're used below.

Edit: Plus all the `var_name_` columns.

Reply:
That makes sense, I added stricter validation in f5ee577a4ba516eb23e9b45a0e244a39af79c7c3. Note that we check for _any_ column matching the pattern `var_name_*`, since technically the function doesn't care what the names are, or how many of them there are.